spdystream test is flaky (on Mac?)

I see unreliable (timing related?) execution of `pkg/util/httpstream/spdy` tests.  In some cases they hang and timeout.  It may be Mac related, since I can reproduce it consistently on my Mac on go 1.4.

```
panic: test timed out after 10s

goroutine 71 [running]:
testing.func·008()
	/usr/local/Cellar/go/1.4.1/libexec/src/testing/testing.go:681 +0x12f
created by time.goFunc
	/usr/local/Cellar/go/1.4.1/libexec/src/time/sleep.go:129 +0x4b

goroutine 1 [chan receive]:
testing.RunTests(0x3fefd0, 0x4fde00, 0x2, 0x2, 0x505501)
	/usr/local/Cellar/go/1.4.1/libexec/src/testing/testing.go:556 +0xad6
testing.(*M).Run(0xc20802e2d0, 0x50c8e0)
	/usr/local/Cellar/go/1.4.1/libexec/src/testing/testing.go:485 +0x6c
main.main()
	github.com/GoogleCloudPlatform/kubernetes/pkg/util/httpstream/spdy/_test/_testmain.go:54 +0x1d5

goroutine 5 [chan receive]:
github.com/golang/glog.(*loggingT).flushDaemon(0x504d00)
	/Users/clayton/projects/go/src/github.com/GoogleCloudPlatform/kubernetes/Godeps/_workspace/src/github.com/golang/glog/glog.go:879 +0x78
created by github.com/golang/glog.init·1
	/Users/clayton/projects/go/src/github.com/GoogleCloudPlatform/kubernetes/Godeps/_workspace/src/github.com/golang/glog/glog.go:410 +0x2a7

goroutine 17 [syscall, locked to thread]:
runtime.goexit()
	/usr/local/Cellar/go/1.4.1/libexec/src/runtime/asm_amd64.s:2232 +0x1

goroutine 7 [semacquire]:
sync.(*WaitGroup).Wait(0xc20802e7b0)
	/usr/local/Cellar/go/1.4.1/libexec/src/sync/waitgroup.go:132 +0x169
net/http/httptest.(*Server).Close(0xc20802e780)
	/usr/local/Cellar/go/1.4.1/libexec/src/net/http/httptest/server.go:168 +0x5f
github.com/GoogleCloudPlatform/kubernetes/pkg/util/httpstream/spdy.TestRoundTripAndNewConnection(0xc208042090)
	/Users/clayton/projects/go/src/github.com/GoogleCloudPlatform/kubernetes/_output/local/go/src/github.com/GoogleCloudPlatform/kubernetes/pkg/util/httpstream/spdy/roundtripper_test.go:226 +0x22fc
testing.tRunner(0xc208042090, 0x4fde00)
	/usr/local/Cellar/go/1.4.1/libexec/src/testing/testing.go:447 +0xbf
created by testing.RunTests
	/usr/local/Cellar/go/1.4.1/libexec/src/testing/testing.go:555 +0xa8b

goroutine 8 [IO wait]:
net.(*pollDesc).Wait(0xc2080101b0, 0x72, 0x0, 0x0)
	/usr/local/Cellar/go/1.4.1/libexec/src/net/fd_poll_runtime.go:84 +0x47
net.(*pollDesc).WaitRead(0xc2080101b0, 0x0, 0x0)
	/usr/local/Cellar/go/1.4.1/libexec/src/net/fd_poll_runtime.go:89 +0x43
net.(*netFD).accept(0xc208010150, 0x0, 0x5bff00, 0xc20802aaf8)
	/usr/local/Cellar/go/1.4.1/libexec/src/net/fd_unix.go:419 +0x40b
net.(*TCPListener).AcceptTCP(0xc20802c058, 0x5c1558, 0x0, 0x0)
	/usr/local/Cellar/go/1.4.1/libexec/src/net/tcpsock_posix.go:234 +0x4e
net.(*TCPListener).Accept(0xc20802c058, 0x0, 0x0, 0x0, 0x0)
	/usr/local/Cellar/go/1.4.1/libexec/src/net/tcpsock_posix.go:244 +0x4c
net/http/httptest.(*historyListener).Accept(0xc20803ccf0, 0x0, 0x0, 0x0, 0x0)
	/usr/local/Cellar/go/1.4.1/libexec/src/net/http/httptest/server.go:48 +0x7a
net/http.(*Server).Serve(0xc208030300, 0x5c13a0, 0xc20803ccf0, 0x0, 0x0)
	/usr/local/Cellar/go/1.4.1/libexec/src/net/http/server.go:1728 +0x92
created by net/http/httptest.(*Server).Start
	/usr/local/Cellar/go/1.4.1/libexec/src/net/http/httptest/server.go:109 +0x35e

goroutine 29 [select]:
github.com/docker/spdystream.(*idleAwareFramer).monitor(0xc20803d3e0)
	/Users/clayton/projects/go/src/github.com/GoogleCloudPlatform/kubernetes/Godeps/_workspace/src/github.com/docker/spdystream/connection.go:55 +0x2d6
created by github.com/docker/spdystream.NewConnection
	/Users/clayton/projects/go/src/github.com/GoogleCloudPlatform/kubernetes/Godeps/_workspace/src/github.com/docker/spdystream/connection.go:184 +0x59b

goroutine 10 [IO wait]:
net.(*pollDesc).Wait(0xc208010530, 0x72, 0x0, 0x0)
	/usr/local/Cellar/go/1.4.1/libexec/src/net/fd_poll_runtime.go:84 +0x47
net.(*pollDesc).WaitRead(0xc208010530, 0x0, 0x0)
	/usr/local/Cellar/go/1.4.1/libexec/src/net/fd_poll_runtime.go:89 +0x43
net.(*netFD).accept(0xc2080104d0, 0x0, 0x5bff00, 0xc20802adb0)
	/usr/local/Cellar/go/1.4.1/libexec/src/net/fd_unix.go:419 +0x40b
net.(*TCPListener).AcceptTCP(0xc20802c080, 0x5c1558, 0x0, 0x0)
	/usr/local/Cellar/go/1.4.1/libexec/src/net/tcpsock_posix.go:234 +0x4e
net.(*TCPListener).Accept(0xc20802c080, 0x0, 0x0, 0x0, 0x0)
	/usr/local/Cellar/go/1.4.1/libexec/src/net/tcpsock_posix.go:244 +0x4c
net/http/httptest.(*historyListener).Accept(0xc20803d110, 0x0, 0x0, 0x0, 0x0)
	/usr/local/Cellar/go/1.4.1/libexec/src/net/http/httptest/server.go:48 +0x7a
net/http.(*Server).Serve(0xc208030420, 0x5c13a0, 0xc20803d110, 0x0, 0x0)
	/usr/local/Cellar/go/1.4.1/libexec/src/net/http/server.go:1728 +0x92
created by net/http/httptest.(*Server).Start
	/usr/local/Cellar/go/1.4.1/libexec/src/net/http/httptest/server.go:109 +0x35e

goroutine 30 [IO wait]:
net.(*pollDesc).Wait(0xc208010d80, 0x72, 0x0, 0x0)
	/usr/local/Cellar/go/1.4.1/libexec/src/net/fd_poll_runtime.go:84 +0x47
net.(*pollDesc).WaitRead(0xc208010d80, 0x0, 0x0)
	/usr/local/Cellar/go/1.4.1/libexec/src/net/fd_poll_runtime.go:89 +0x43
net.(*netFD).Read(0xc208010d20, 0xc20802ae30, 0x4, 0x8, 0x0, 0x5bff00, 0xc20802ae38)
	/usr/local/Cellar/go/1.4.1/libexec/src/net/fd_unix.go:242 +0x40f
net.(*conn).Read(0xc20802c020, 0xc20802ae30, 0x4, 0x8, 0x0, 0x0, 0x0)
	/usr/local/Cellar/go/1.4.1/libexec/src/net/net.go:121 +0xdc
io.ReadAtLeast(0x5c14e0, 0xc20802c020, 0xc20802ae30, 0x4, 0x8, 0x4, 0x0, 0x0, 0x0)
	/usr/local/Cellar/go/1.4.1/libexec/src/io/io.go:298 +0xf1
io.ReadFull(0x5c14e0, 0xc20802c020, 0xc20802ae30, 0x4, 0x8, 0xf3924, 0x0, 0x0)
	/usr/local/Cellar/go/1.4.1/libexec/src/io/io.go:316 +0x6d
encoding/binary.Read(0x5c14e0, 0xc20802c020, 0x5c28c0, 0x50d2e0, 0x266ec0, 0xc20802ae28, 0x0, 0x0)
	/usr/local/Cellar/go/1.4.1/libexec/src/encoding/binary/binary.go:148 +0x12d
code.google.com/p/go.net/spdy.(*Framer).ReadFrame(0xc208030900, 0x0, 0x0, 0x0, 0x0)
	/Users/clayton/projects/go/src/github.com/GoogleCloudPlatform/kubernetes/Godeps/_workspace/src/code.google.com/p/go.net/spdy/read.go:148 +0xff
github.com/docker/spdystream.(*idleAwareFramer).ReadFrame(0xc20803d3e0, 0x0, 0x0, 0x0, 0x0)
	/Users/clayton/projects/go/src/github.com/GoogleCloudPlatform/kubernetes/Godeps/_workspace/src/github.com/docker/spdystream/connection.go:101 +0x4f
github.com/docker/spdystream.(*Connection).Serve(0xc2080404d0, 0xc20802ad00)
	/Users/clayton/projects/go/src/github.com/GoogleCloudPlatform/kubernetes/Godeps/_workspace/src/github.com/docker/spdystream/connection.go:244 +0x206
created by github.com/GoogleCloudPlatform/kubernetes/pkg/util/httpstream/spdy.newConnection
	/Users/clayton/projects/go/src/github.com/GoogleCloudPlatform/kubernetes/_output/local/go/src/github.com/GoogleCloudPlatform/kubernetes/pkg/util/httpstream/spdy/connection.go:68 +0x113

goroutine 12 [IO wait]:
net.(*pollDesc).Wait(0xc2080108b0, 0x72, 0x0, 0x0)
	/usr/local/Cellar/go/1.4.1/libexec/src/net/fd_poll_runtime.go:84 +0x47
net.(*pollDesc).WaitRead(0xc2080108b0, 0x0, 0x0)
	/usr/local/Cellar/go/1.4.1/libexec/src/net/fd_poll_runtime.go:89 +0x43
net.(*netFD).accept(0xc208010850, 0x0, 0x5bff00, 0xc20802afa0)
	/usr/local/Cellar/go/1.4.1/libexec/src/net/fd_unix.go:419 +0x40b
net.(*TCPListener).AcceptTCP(0xc20802c0a0, 0x5c1558, 0x0, 0x0)
	/usr/local/Cellar/go/1.4.1/libexec/src/net/tcpsock_posix.go:234 +0x4e
net.(*TCPListener).Accept(0xc20802c0a0, 0x0, 0x0, 0x0, 0x0)
	/usr/local/Cellar/go/1.4.1/libexec/src/net/tcpsock_posix.go:244 +0x4c
net/http/httptest.(*historyListener).Accept(0xc20803d4d0, 0x0, 0x0, 0x0, 0x0)
	/usr/local/Cellar/go/1.4.1/libexec/src/net/http/httptest/server.go:48 +0x7a
net/http.(*Server).Serve(0xc208030540, 0x5c13a0, 0xc20803d4d0, 0x0, 0x0)
	/usr/local/Cellar/go/1.4.1/libexec/src/net/http/server.go:1728 +0x92
created by net/http/httptest.(*Server).Start
	/usr/local/Cellar/go/1.4.1/libexec/src/net/http/httptest/server.go:109 +0x35e

goroutine 31 [chan receive]:
github.com/docker/spdystream.func·001(0xc20803cf00)
	/Users/clayton/projects/go/src/github.com/GoogleCloudPlatform/kubernetes/Godeps/_workspace/src/github.com/docker/spdystream/connection.go:235 +0x43
created by github.com/docker/spdystream.(*Connection).Serve
	/Users/clayton/projects/go/src/github.com/GoogleCloudPlatform/kubernetes/Godeps/_workspace/src/github.com/docker/spdystream/connection.go:237 +0x175

goroutine 14 [IO wait]:
net.(*pollDesc).Wait(0xc208010c30, 0x72, 0x0, 0x0)
	/usr/local/Cellar/go/1.4.1/libexec/src/net/fd_poll_runtime.go:84 +0x47
net.(*pollDesc).WaitRead(0xc208010c30, 0x0, 0x0)
	/usr/local/Cellar/go/1.4.1/libexec/src/net/fd_poll_runtime.go:89 +0x43
net.(*netFD).accept(0xc208010bd0, 0x0, 0x5bff00, 0xc20802aa40)
	/usr/local/Cellar/go/1.4.1/libexec/src/net/fd_unix.go:419 +0x40b
net.(*TCPListener).AcceptTCP(0xc20802c0c0, 0x5c1558, 0x0, 0x0)
	/usr/local/Cellar/go/1.4.1/libexec/src/net/tcpsock_posix.go:234 +0x4e
net.(*TCPListener).Accept(0xc20802c0c0, 0x0, 0x0, 0x0, 0x0)
	/usr/local/Cellar/go/1.4.1/libexec/src/net/tcpsock_posix.go:244 +0x4c
net/http/httptest.(*historyListener).Accept(0xc20803d890, 0x0, 0x0, 0x0, 0x0)
	/usr/local/Cellar/go/1.4.1/libexec/src/net/http/httptest/server.go:48 +0x7a
net/http.(*Server).Serve(0xc208030660, 0x5c13a0, 0xc20803d890, 0x0, 0x0)
	/usr/local/Cellar/go/1.4.1/libexec/src/net/http/server.go:1728 +0x92
created by net/http/httptest.(*Server).Start
	/usr/local/Cellar/go/1.4.1/libexec/src/net/http/httptest/server.go:109 +0x35e

goroutine 15 [select]:
github.com/docker/spdystream.(*Stream).Read(0xc20806c780, 0xc20838c000, 0x8000, 0x8000, 0x5, 0x0, 0x0)
	/Users/clayton/projects/go/src/github.com/GoogleCloudPlatform/kubernetes/Godeps/_workspace/src/github.com/docker/spdystream/stream.go:82 +0x347
io.Copy(0x5c2ee0, 0xc20806c780, 0x5c2f08, 0xc20806c780, 0x5, 0x0, 0x0)
	/usr/local/Cellar/go/1.4.1/libexec/src/io/io.go:362 +0x1f6
github.com/GoogleCloudPlatform/kubernetes/pkg/util/httpstream/spdy.func·002(0x5c16b0, 0xc20806c500, 0xc20801e340)
	/Users/clayton/projects/go/src/github.com/GoogleCloudPlatform/kubernetes/_output/local/go/src/github.com/GoogleCloudPlatform/kubernetes/pkg/util/httpstream/spdy/roundtripper_test.go:100 +0x6d3
net/http.HandlerFunc.ServeHTTP(0xc20800abe0, 0x5c16b0, 0xc20806c500, 0xc20801e340)
	/usr/local/Cellar/go/1.4.1/libexec/src/net/http/server.go:1265 +0x41
net/http/httptest.(*waitGroupHandler).ServeHTTP(0xc20800ac60, 0x5c16b0, 0xc20806c500, 0xc20801e340)
	/usr/local/Cellar/go/1.4.1/libexec/src/net/http/httptest/server.go:200 +0xb7
net/http.serverHandler.ServeHTTP(0xc208030660, 0x5c16b0, 0xc20806c500, 0xc20801e340)
	/usr/local/Cellar/go/1.4.1/libexec/src/net/http/server.go:1703 +0x19a
net/http.(*conn).serve(0xc20806c820)
	/usr/local/Cellar/go/1.4.1/libexec/src/net/http/server.go:1204 +0xb57
created by net/http.(*Server).Serve
	/usr/local/Cellar/go/1.4.1/libexec/src/net/http/server.go:1751 +0x35e

goroutine 16 [select]:
github.com/docker/spdystream.(*idleAwareFramer).monitor(0xc20803cff0)
	/Users/clayton/projects/go/src/github.com/GoogleCloudPlatform/kubernetes/Godeps/_workspace/src/github.com/docker/spdystream/connection.go:55 +0x2d6
created by github.com/docker/spdystream.NewConnection
	/Users/clayton/projects/go/src/github.com/GoogleCloudPlatform/kubernetes/Godeps/_workspace/src/github.com/docker/spdystream/connection.go:184 +0x59b

goroutine 18 [IO wait]:
net.(*pollDesc).Wait(0xc208010df0, 0x72, 0x0, 0x0)
	/usr/local/Cellar/go/1.4.1/libexec/src/net/fd_poll_runtime.go:84 +0x47
net.(*pollDesc).WaitRead(0xc208010df0, 0x0, 0x0)
	/usr/local/Cellar/go/1.4.1/libexec/src/net/fd_poll_runtime.go:89 +0x43
net.(*netFD).Read(0xc208010d90, 0xc20802b168, 0x4, 0x8, 0x0, 0x5bff00, 0xc20802add0)
	/usr/local/Cellar/go/1.4.1/libexec/src/net/fd_unix.go:242 +0x40f
net.(*conn).Read(0xc20802c0d0, 0xc20802b168, 0x4, 0x8, 0x0, 0x0, 0x0)
	/usr/local/Cellar/go/1.4.1/libexec/src/net/net.go:121 +0xdc
io.ReadAtLeast(0x5c14e0, 0xc20802c0d0, 0xc20802b168, 0x4, 0x8, 0x4, 0x0, 0x0, 0x0)
	/usr/local/Cellar/go/1.4.1/libexec/src/io/io.go:298 +0xf1
io.ReadFull(0x5c14e0, 0xc20802c0d0, 0xc20802b168, 0x4, 0x8, 0xf3924, 0x0, 0x0)
	/usr/local/Cellar/go/1.4.1/libexec/src/io/io.go:316 +0x6d
encoding/binary.Read(0x5c14e0, 0xc20802c0d0, 0x5c28c0, 0x50d2e0, 0x266ec0, 0xc20802b160, 0x0, 0x0)
	/usr/local/Cellar/go/1.4.1/libexec/src/encoding/binary/binary.go:148 +0x12d
code.google.com/p/go.net/spdy.(*Framer).ReadFrame(0xc208030600, 0x0, 0x0, 0x0, 0x0)
	/Users/clayton/projects/go/src/github.com/GoogleCloudPlatform/kubernetes/Godeps/_workspace/src/code.google.com/p/go.net/spdy/read.go:148 +0xff
github.com/docker/spdystream.(*idleAwareFramer).ReadFrame(0xc20803cff0, 0x0, 0x0, 0x0, 0x0)
	/Users/clayton/projects/go/src/github.com/GoogleCloudPlatform/kubernetes/Godeps/_workspace/src/github.com/docker/spdystream/connection.go:101 +0x4f
github.com/docker/spdystream.(*Connection).Serve(0xc2080402c0, 0xc20802abd0)
	/Users/clayton/projects/go/src/github.com/GoogleCloudPlatform/kubernetes/Godeps/_workspace/src/github.com/docker/spdystream/connection.go:244 +0x206
created by github.com/GoogleCloudPlatform/kubernetes/pkg/util/httpstream/spdy.newConnection
	/Users/clayton/projects/go/src/github.com/GoogleCloudPlatform/kubernetes/_output/local/go/src/github.com/GoogleCloudPlatform/kubernetes/pkg/util/httpstream/spdy/connection.go:68 +0x113

goroutine 19 [chan receive]:
github.com/docker/spdystream.func·001(0xc20803d0b0)
	/Users/clayton/projects/go/src/github.com/GoogleCloudPlatform/kubernetes/Godeps/_workspace/src/github.com/docker/spdystream/connection.go:235 +0x43
created by github.com/docker/spdystream.(*Connection).Serve
	/Users/clayton/projects/go/src/github.com/GoogleCloudPlatform/kubernetes/Godeps/_workspace/src/github.com/docker/spdystream/connection.go:237 +0x175

goroutine 20 [semacquire]:
sync.(*Cond).Wait(0xc20803aec0)
	/usr/local/Cellar/go/1.4.1/libexec/src/sync/cond.go:62 +0x9e
github.com/docker/spdystream.(*PriorityFrameQueue).Pop(0xc20803d0b0, 0x0, 0x0)
	/Users/clayton/projects/go/src/github.com/GoogleCloudPlatform/kubernetes/Godeps/_workspace/src/github.com/docker/spdystream/priority.go:85 +0x11c
github.com/docker/spdystream.(*Connection).frameHandler(0xc2080402c0, 0xc20803d0b0, 0xc20802abd0)
	/Users/clayton/projects/go/src/github.com/GoogleCloudPlatform/kubernetes/Godeps/_workspace/src/github.com/docker/spdystream/connection.go:308 +0x33
created by github.com/docker/spdystream.(*Connection).Serve
	/Users/clayton/projects/go/src/github.com/GoogleCloudPlatform/kubernetes/Godeps/_workspace/src/github.com/docker/spdystream/connection.go:239 +0x1cc

goroutine 21 [chan receive]:
github.com/docker/spdystream.func·001(0xc20803d140)
	/Users/clayton/projects/go/src/github.com/GoogleCloudPlatform/kubernetes/Godeps/_workspace/src/github.com/docker/spdystream/connection.go:235 +0x43
created by github.com/docker/spdystream.(*Connection).Serve
	/Users/clayton/projects/go/src/github.com/GoogleCloudPlatform/kubernetes/Godeps/_workspace/src/github.com/docker/spdystream/connection.go:237 +0x175

goroutine 22 [semacquire]:
sync.(*Cond).Wait(0xc20803af00)
	/usr/local/Cellar/go/1.4.1/libexec/src/sync/cond.go:62 +0x9e
github.com/docker/spdystream.(*PriorityFrameQueue).Pop(0xc20803d140, 0x0, 0x0)
	/Users/clayton/projects/go/src/github.com/GoogleCloudPlatform/kubernetes/Godeps/_workspace/src/github.com/docker/spdystream/priority.go:85 +0x11c
github.com/docker/spdystream.(*Connection).frameHandler(0xc2080402c0, 0xc20803d140, 0xc20802abd0)
	/Users/clayton/projects/go/src/github.com/GoogleCloudPlatform/kubernetes/Godeps/_workspace/src/github.com/docker/spdystream/connection.go:308 +0x33
created by github.com/docker/spdystream.(*Connection).Serve
	/Users/clayton/projects/go/src/github.com/GoogleCloudPlatform/kubernetes/Godeps/_workspace/src/github.com/docker/spdystream/connection.go:239 +0x1cc

goroutine 23 [chan receive]:
github.com/docker/spdystream.func·001(0xc20803d170)
	/Users/clayton/projects/go/src/github.com/GoogleCloudPlatform/kubernetes/Godeps/_workspace/src/github.com/docker/spdystream/connection.go:235 +0x43
created by github.com/docker/spdystream.(*Connection).Serve
	/Users/clayton/projects/go/src/github.com/GoogleCloudPlatform/kubernetes/Godeps/_workspace/src/github.com/docker/spdystream/connection.go:237 +0x175

goroutine 24 [semacquire]:
sync.(*Cond).Wait(0xc20803af40)
	/usr/local/Cellar/go/1.4.1/libexec/src/sync/cond.go:62 +0x9e
github.com/docker/spdystream.(*PriorityFrameQueue).Pop(0xc20803d170, 0x0, 0x0)
	/Users/clayton/projects/go/src/github.com/GoogleCloudPlatform/kubernetes/Godeps/_workspace/src/github.com/docker/spdystream/priority.go:85 +0x11c
github.com/docker/spdystream.(*Connection).frameHandler(0xc2080402c0, 0xc20803d170, 0xc20802abd0)
	/Users/clayton/projects/go/src/github.com/GoogleCloudPlatform/kubernetes/Godeps/_workspace/src/github.com/docker/spdystream/connection.go:308 +0x33
created by github.com/docker/spdystream.(*Connection).Serve
	/Users/clayton/projects/go/src/github.com/GoogleCloudPlatform/kubernetes/Godeps/_workspace/src/github.com/docker/spdystream/connection.go:239 +0x1cc

goroutine 25 [chan receive]:
github.com/docker/spdystream.func·001(0xc20803d1a0)
	/Users/clayton/projects/go/src/github.com/GoogleCloudPlatform/kubernetes/Godeps/_workspace/src/github.com/docker/spdystream/connection.go:235 +0x43
created by github.com/docker/spdystream.(*Connection).Serve
	/Users/clayton/projects/go/src/github.com/GoogleCloudPlatform/kubernetes/Godeps/_workspace/src/github.com/docker/spdystream/connection.go:237 +0x175

goroutine 26 [semacquire]:
sync.(*Cond).Wait(0xc20803af80)
	/usr/local/Cellar/go/1.4.1/libexec/src/sync/cond.go:62 +0x9e
github.com/docker/spdystream.(*PriorityFrameQueue).Pop(0xc20803d1a0, 0x0, 0x0)
	/Users/clayton/projects/go/src/github.com/GoogleCloudPlatform/kubernetes/Godeps/_workspace/src/github.com/docker/spdystream/priority.go:85 +0x11c
github.com/docker/spdystream.(*Connection).frameHandler(0xc2080402c0, 0xc20803d1a0, 0xc20802abd0)
	/Users/clayton/projects/go/src/github.com/GoogleCloudPlatform/kubernetes/Godeps/_workspace/src/github.com/docker/spdystream/connection.go:308 +0x33
created by github.com/docker/spdystream.(*Connection).Serve
	/Users/clayton/projects/go/src/github.com/GoogleCloudPlatform/kubernetes/Godeps/_workspace/src/github.com/docker/spdystream/connection.go:239 +0x1cc

goroutine 27 [chan receive]:
github.com/docker/spdystream.func·001(0xc20803d230)
	/Users/clayton/projects/go/src/github.com/GoogleCloudPlatform/kubernetes/Godeps/_workspace/src/github.com/docker/spdystream/connection.go:235 +0x43
created by github.com/docker/spdystream.(*Connection).Serve
	/Users/clayton/projects/go/src/github.com/GoogleCloudPlatform/kubernetes/Godeps/_workspace/src/github.com/docker/spdystream/connection.go:237 +0x175

goroutine 28 [semacquire]:
sync.(*Cond).Wait(0xc20803afc0)
	/usr/local/Cellar/go/1.4.1/libexec/src/sync/cond.go:62 +0x9e
github.com/docker/spdystream.(*PriorityFrameQueue).Pop(0xc20803d230, 0x0, 0x0)
	/Users/clayton/projects/go/src/github.com/GoogleCloudPlatform/kubernetes/Godeps/_workspace/src/github.com/docker/spdystream/priority.go:85 +0x11c
github.com/docker/spdystream.(*Connection).frameHandler(0xc2080402c0, 0xc20803d230, 0xc20802abd0)
	/Users/clayton/projects/go/src/github.com/GoogleCloudPlatform/kubernetes/Godeps/_workspace/src/github.com/docker/spdystream/connection.go:308 +0x33
created by github.com/docker/spdystream.(*Connection).Serve
	/Users/clayton/projects/go/src/github.com/GoogleCloudPlatform/kubernetes/Godeps/_workspace/src/github.com/docker/spdystream/connection.go:239 +0x1cc

goroutine 32 [semacquire]:
sync.(*Cond).Wait(0xc20803ad40)
	/usr/local/Cellar/go/1.4.1/libexec/src/sync/cond.go:62 +0x9e
github.com/docker/spdystream.(*PriorityFrameQueue).Pop(0xc20803cf00, 0x0, 0x0)
	/Users/clayton/projects/go/src/github.com/GoogleCloudPlatform/kubernetes/Godeps/_workspace/src/github.com/docker/spdystream/priority.go:85 +0x11c
github.com/docker/spdystream.(*Connection).frameHandler(0xc2080404d0, 0xc20803cf00, 0xc20802ad00)
	/Users/clayton/projects/go/src/github.com/GoogleCloudPlatform/kubernetes/Godeps/_workspace/src/github.com/docker/spdystream/connection.go:308 +0x33
created by github.com/docker/spdystream.(*Connection).Serve
	/Users/clayton/projects/go/src/github.com/GoogleCloudPlatform/kubernetes/Godeps/_workspace/src/github.com/docker/spdystream/connection.go:239 +0x1cc

goroutine 33 [chan receive]:
github.com/docker/spdystream.func·001(0xc20803d1d0)
	/Users/clayton/projects/go/src/github.com/GoogleCloudPlatform/kubernetes/Godeps/_workspace/src/github.com/docker/spdystream/connection.go:235 +0x43
created by github.com/docker/spdystream.(*Connection).Serve
	/Users/clayton/projects/go/src/github.com/GoogleCloudPlatform/kubernetes/Godeps/_workspace/src/github.com/docker/spdystream/connection.go:237 +0x175

goroutine 34 [semacquire]:
sync.(*Cond).Wait(0xc20803ae80)
	/usr/local/Cellar/go/1.4.1/libexec/src/sync/cond.go:62 +0x9e
github.com/docker/spdystream.(*PriorityFrameQueue).Pop(0xc20803d1d0, 0x0, 0x0)
	/Users/clayton/projects/go/src/github.com/GoogleCloudPlatform/kubernetes/Godeps/_workspace/src/github.com/docker/spdystream/priority.go:85 +0x11c
github.com/docker/spdystream.(*Connection).frameHandler(0xc2080404d0, 0xc20803d1d0, 0xc20802ad00)
	/Users/clayton/projects/go/src/github.com/GoogleCloudPlatform/kubernetes/Godeps/_workspace/src/github.com/docker/spdystream/connection.go:308 +0x33
created by github.com/docker/spdystream.(*Connection).Serve
	/Users/clayton/projects/go/src/github.com/GoogleCloudPlatform/kubernetes/Godeps/_workspace/src/github.com/docker/spdystream/connection.go:239 +0x1cc

goroutine 35 [chan receive]:
github.com/docker/spdystream.func·001(0xc20803d200)
	/Users/clayton/projects/go/src/github.com/GoogleCloudPlatform/kubernetes/Godeps/_workspace/src/github.com/docker/spdystream/connection.go:235 +0x43
created by github.com/docker/spdystream.(*Connection).Serve
	/Users/clayton/projects/go/src/github.com/GoogleCloudPlatform/kubernetes/Godeps/_workspace/src/github.com/docker/spdystream/connection.go:237 +0x175

goroutine 36 [semacquire]:
sync.(*Cond).Wait(0xc20803b3c0)
	/usr/local/Cellar/go/1.4.1/libexec/src/sync/cond.go:62 +0x9e
github.com/docker/spdystream.(*PriorityFrameQueue).Pop(0xc20803d200, 0x0, 0x0)
	/Users/clayton/projects/go/src/github.com/GoogleCloudPlatform/kubernetes/Godeps/_workspace/src/github.com/docker/spdystream/priority.go:85 +0x11c
github.com/docker/spdystream.(*Connection).frameHandler(0xc2080404d0, 0xc20803d200, 0xc20802ad00)
	/Users/clayton/projects/go/src/github.com/GoogleCloudPlatform/kubernetes/Godeps/_workspace/src/github.com/docker/spdystream/connection.go:308 +0x33
created by github.com/docker/spdystream.(*Connection).Serve
	/Users/clayton/projects/go/src/github.com/GoogleCloudPlatform/kubernetes/Godeps/_workspace/src/github.com/docker/spdystream/connection.go:239 +0x1cc

goroutine 37 [chan receive]:
github.com/docker/spdystream.func·001(0xc20803d2c0)
	/Users/clayton/projects/go/src/github.com/GoogleCloudPlatform/kubernetes/Godeps/_workspace/src/github.com/docker/spdystream/connection.go:235 +0x43
created by github.com/docker/spdystream.(*Connection).Serve
	/Users/clayton/projects/go/src/github.com/GoogleCloudPlatform/kubernetes/Godeps/_workspace/src/github.com/docker/spdystream/connection.go:237 +0x175

goroutine 38 [semacquire]:
sync.(*Cond).Wait(0xc20803b480)
	/usr/local/Cellar/go/1.4.1/libexec/src/sync/cond.go:62 +0x9e
github.com/docker/spdystream.(*PriorityFrameQueue).Pop(0xc20803d2c0, 0x0, 0x0)
	/Users/clayton/projects/go/src/github.com/GoogleCloudPlatform/kubernetes/Godeps/_workspace/src/github.com/docker/spdystream/priority.go:85 +0x11c
github.com/docker/spdystream.(*Connection).frameHandler(0xc2080404d0, 0xc20803d2c0, 0xc20802ad00)
	/Users/clayton/projects/go/src/github.com/GoogleCloudPlatform/kubernetes/Godeps/_workspace/src/github.com/docker/spdystream/connection.go:308 +0x33
created by github.com/docker/spdystream.(*Connection).Serve
	/Users/clayton/projects/go/src/github.com/GoogleCloudPlatform/kubernetes/Godeps/_workspace/src/github.com/docker/spdystream/connection.go:239 +0x1cc

goroutine 39 [chan receive]:
github.com/docker/spdystream.func·001(0xc20803d2f0)
	/Users/clayton/projects/go/src/github.com/GoogleCloudPlatform/kubernetes/Godeps/_workspace/src/github.com/docker/spdystream/connection.go:235 +0x43
created by github.com/docker/spdystream.(*Connection).Serve
	/Users/clayton/projects/go/src/github.com/GoogleCloudPlatform/kubernetes/Godeps/_workspace/src/github.com/docker/spdystream/connection.go:237 +0x175

goroutine 40 [semacquire]:
sync.(*Cond).Wait(0xc20803b4c0)
	/usr/local/Cellar/go/1.4.1/libexec/src/sync/cond.go:62 +0x9e
github.com/docker/spdystream.(*PriorityFrameQueue).Pop(0xc20803d2f0, 0x0, 0x0)
	/Users/clayton/projects/go/src/github.com/GoogleCloudPlatform/kubernetes/Godeps/_workspace/src/github.com/docker/spdystream/priority.go:85 +0x11c
github.com/docker/spdystream.(*Connection).frameHandler(0xc2080404d0, 0xc20803d2f0, 0xc20802ad00)
	/Users/clayton/projects/go/src/github.com/GoogleCloudPlatform/kubernetes/Godeps/_workspace/src/github.com/docker/spdystream/connection.go:308 +0x33
created by github.com/docker/spdystream.(*Connection).Serve
	/Users/clayton/projects/go/src/github.com/GoogleCloudPlatform/kubernetes/Godeps/_workspace/src/github.com/docker/spdystream/connection.go:239 +0x1cc

goroutine 42 [chan send]:
github.com/docker/spdystream.(*idleAwareFramer).WriteFrame(0xc2080744e0, 0x5cccd0, 0xc20800bb20, 0x0, 0x0)
	/Users/clayton/projects/go/src/github.com/GoogleCloudPlatform/kubernetes/Godeps/_workspace/src/github.com/docker/spdystream/connection.go:95 +0xa7
github.com/docker/spdystream.(*Stream).Reset(0xc20809a1e0, 0x0, 0x0)
	/Users/clayton/projects/go/src/github.com/GoogleCloudPlatform/kubernetes/Godeps/_workspace/src/github.com/docker/spdystream/stream.go:188 +0x1b8
github.com/GoogleCloudPlatform/kubernetes/pkg/util/httpstream/spdy.(*connection).Close(0xc208074570, 0x0, 0x0)
	/Users/clayton/projects/go/src/github.com/GoogleCloudPlatform/kubernetes/_output/local/go/src/github.com/GoogleCloudPlatform/kubernetes/pkg/util/httpstream/spdy/connection.go:81 +0xd0
github.com/GoogleCloudPlatform/kubernetes/pkg/util/httpstream/spdy.func·002(0x5c16b0, 0xc2085b6e60, 0xc2085989c0)
	/Users/clayton/projects/go/src/github.com/GoogleCloudPlatform/kubernetes/_output/local/go/src/github.com/GoogleCloudPlatform/kubernetes/pkg/util/httpstream/spdy/roundtripper_test.go:101 +0x6d4
net/http.HandlerFunc.ServeHTTP(0xc20800b200, 0x5c16b0, 0xc2085b6e60, 0xc2085989c0)
	/usr/local/Cellar/go/1.4.1/libexec/src/net/http/server.go:1265 +0x41
net/http/httptest.(*waitGroupHandler).ServeHTTP(0xc208472a40, 0x5c16b0, 0xc2085b6e60, 0xc2085989c0)
	/usr/local/Cellar/go/1.4.1/libexec/src/net/http/httptest/server.go:200 +0xb7
net/http.serverHandler.ServeHTTP(0xc208030d20, 0x5c16b0, 0xc2085b6e60, 0xc2085989c0)
	/usr/local/Cellar/go/1.4.1/libexec/src/net/http/server.go:1703 +0x19a
net/http.(*conn).serve(0xc208500000)
	/usr/local/Cellar/go/1.4.1/libexec/src/net/http/server.go:1204 +0xb57
created by net/http.(*Server).Serve
	/usr/local/Cellar/go/1.4.1/libexec/src/net/http/server.go:1751 +0x35e
FAIL	github.com/GoogleCloudPlatform/kubernetes/pkg/util/httpstream/spdy	10.014s
```

--- a/Godeps/Godeps.json
+++ b/Godeps/Godeps.json
@@ -114,7 +114,7 @@
 		},
 		{
 			"ImportPath": "github.com/docker/spdystream",
-			"Rev": "e9bf9912b85eec0ed6aaf317808a0eab25e3ca43"
+			"Rev": "e731c8f9f19ffd7e51a469a2de1580c1dfbb4fae"
 		},
 		{
 			"ImportPath": "github.com/elazarl/go-bindata-assetfs",

--- a/Godeps/_workspace/src/github.com/docker/spdystream/connection.go
+++ b/Godeps/_workspace/src/github.com/docker/spdystream/connection.go
@@ -31,6 +31,7 @@ type AuthHandler func(header http.Header, slot uint8, parent uint32) bool
 type idleAwareFramer struct {
 	f              *spdy.Framer
 	conn           *Connection
+	writeLock      sync.Mutex
 	resetChan      chan struct{}
 	setTimeoutChan chan time.Duration
 	timeout        time.Duration
@@ -47,8 +48,9 @@ func newIdleAwareFramer(framer *spdy.Framer) *idleAwareFramer {
 
 func (i *idleAwareFramer) monitor() {
 	var (
-		timer   *time.Timer
-		expired <-chan time.Time
+		timer     *time.Timer
+		expired   <-chan time.Time
+		resetChan = i.resetChan
 	)
 Loop:
 	for {
@@ -67,7 +69,7 @@ Loop:
 					timer.Reset(timeout)
 				}
 			}
-		case <-i.resetChan:
+		case <-resetChan:
 			if timer != nil && i.timeout > 0 {
 				timer.Reset(i.timeout)
 			}
@@ -87,12 +89,25 @@ Loop:
 			if timer != nil {
 				timer.Stop()
 			}
+			i.writeLock.Lock()
+			close(resetChan)
+			i.resetChan = nil
+			i.writeLock.Unlock()
 			break Loop
 		}
+	}
+
+	// Drain resetChan
+	for _ = range resetChan {
 	}
 }
 
 func (i *idleAwareFramer) WriteFrame(frame spdy.Frame) error {
+	i.writeLock.Lock()
+	defer i.writeLock.Unlock()
+	if i.resetChan == nil {
+		return io.EOF
+	}
 	err := i.f.WriteFrame(frame)
 	if err != nil {
 		return err
@@ -109,15 +124,18 @@ func (i *idleAwareFramer) ReadFrame() (spdy.Frame, error) {
 		return nil, err
 	}
 
+	// resetChan should never be closed since it is only closed
+	// when the connection has closed its closeChan. This closure
+	// only occurs after all Reads have finished
+	// TODO (dmcgowan): refactor relationship into connection
 	i.resetChan <- struct{}{}
 
 	return frame, nil
 }
 
 type Connection struct {
-	conn      net.Conn
-	framer    *idleAwareFramer
-	writeLock sync.Mutex
+	conn   net.Conn
+	framer *idleAwareFramer
 
 	closeChan      chan bool
 	goneAway       bool
@@ -209,9 +227,7 @@ func (s *Connection) Ping() (time.Duration, error) {
 
 	frame := &spdy.PingFrame{Id: pid}
 	startTime := time.Now()
-	s.writeLock.Lock()
 	writeErr := s.framer.WriteFrame(frame)
-	s.writeLock.Unlock()
 	if writeErr != nil {
 		return time.Duration(0), writeErr
 	}
@@ -512,8 +528,6 @@ func (s *Connection) handleDataFrame(frame *spdy.DataFrame) error {
 
 func (s *Connection) handlePingFrame(frame *spdy.PingFrame) error {
 	if s.pingId&0x01 != frame.Id&0x01 {
-		s.writeLock.Lock()
-		defer s.writeLock.Unlock()
 		return s.framer.WriteFrame(frame)
 	}
 	pingChan, pingOk := s.pingChans[frame.Id]
@@ -663,9 +677,7 @@ func (s *Connection) Close() error {
 		Status:           spdy.GoAwayOK,
 	}
 
-	s.writeLock.Lock()
 	err := s.framer.WriteFrame(goAwayFrame)
-	s.writeLock.Unlock()
 	if err != nil {
 		return err
 	}
@@ -750,8 +762,6 @@ func (s *Connection) sendHeaders(headers http.Header, stream *Stream, fin bool) 
 		CFHeader: spdy.ControlFrameHeader{Flags: flags},
 	}
 
-	s.writeLock.Lock()
-	defer s.writeLock.Unlock()
 	return s.framer.WriteFrame(headerFrame)
 }
 
@@ -767,8 +777,6 @@ func (s *Connection) sendReply(headers http.Header, stream *Stream, fin bool) er
 		CFHeader: spdy.ControlFrameHeader{Flags: flags},
 	}
 
-	s.writeLock.Lock()
-	defer s.writeLock.Unlock()
 	return s.framer.WriteFrame(replyFrame)
 }
 
@@ -778,8 +786,6 @@ func (s *Connection) sendResetFrame(status spdy.RstStreamStatus, streamId spdy.S
 		Status:   status,
 	}
 
-	s.writeLock.Lock()
-	defer s.writeLock.Unlock()
 	return s.framer.WriteFrame(resetFrame)
 }
 
@@ -806,8 +812,6 @@ func (s *Connection) sendStream(stream *Stream, fin bool) error {
 		CFHeader:             spdy.ControlFrameHeader{Flags: flags},
 	}
 
-	s.writeLock.Lock()
-	defer s.writeLock.Unlock()
 	return s.framer.WriteFrame(streamFrame)
 }
 

--- a/Godeps/_workspace/src/github.com/docker/spdystream/spdy_test.go
+++ b/Godeps/_workspace/src/github.com/docker/spdystream/spdy_test.go
@@ -1,10 +1,12 @@
 package spdystream
 
 import (
+	"bufio"
 	"bytes"
 	"io"
 	"net"
 	"net/http"
+	"net/http/httptest"
 	"sync"
 	"testing"
 	"time"
@@ -322,8 +324,6 @@ func TestUnexpectedRemoteConnectionClosed(t *testing.T) {
 			if e == nil || e != io.EOF {
 				t.Fatalf("(%d) Expected to get an EOF stream error", tix)
 			}
-		case <-time.After(500 * time.Millisecond):
-			t.Fatalf("(%d) Timeout waiting for stream closure", tix)
 		}
 
 		closeErr = conn.Close()
@@ -381,8 +381,6 @@ func TestCloseNotification(t *testing.T) {
 	var serverConn net.Conn
 	select {
 	case serverConn = <-serverConnChan:
-	case <-time.After(500 * time.Millisecond):
-		t.Fatal("Timed out waiting for connection closed notification")
 	}
 
 	err = serverConn.Close()
@@ -522,11 +520,7 @@ func TestIdleNoData(t *testing.T) {
 	go spdyConn.Serve(NoOpStreamHandler)
 
 	spdyConn.SetIdleTimeout(10 * time.Millisecond)
-	select {
-	case <-spdyConn.CloseChan():
-	case <-time.After(20 * time.Millisecond):
-		t.Fatal("Timed out waiting for idle connection closure")
-	}
+	<-spdyConn.CloseChan()
 
 	closeErr := server.Close()
 	if closeErr != nil {
@@ -577,8 +571,6 @@ func TestIdleWithData(t *testing.T) {
 
 	writesFinished := false
 
-	expired := time.NewTimer(200 * time.Millisecond)
-
 Loop:
 	for {
 		select {
@@ -589,8 +581,6 @@ Loop:
 				t.Fatal("Connection closed before all writes finished")
 			}
 			break Loop
-		case <-expired.C:
-			t.Fatal("Timed out waiting for idle connection closure")
 		}
 	}
 
@@ -782,6 +772,109 @@ func TestStreamResetWithDataRemaining(t *testing.T) {
 		t.Fatalf("Error shutting down server: %s", closeErr)
 	}
 	wg.Wait()
+}
+
+type roundTripper struct {
+	conn net.Conn
+}
+
+func (s *roundTripper) RoundTrip(req *http.Request) (*http.Response, error) {
+	r := *req
+	req = &r
+
+	conn, err := net.Dial("tcp", req.URL.Host)
+	if err != nil {
+		return nil, err
+	}
+
+	err = req.Write(conn)
+	if err != nil {
+		return nil, err
+	}
+
+	resp, err := http.ReadResponse(bufio.NewReader(conn), req)
+	if err != nil {
+		return nil, err
+	}
+
+	s.conn = conn
+
+	return resp, nil
+}
+
+// see https://github.com/GoogleCloudPlatform/kubernetes/issues/4882
+func TestFramingAfterRemoteConnectionClosed(t *testing.T) {
+	server := httptest.NewUnstartedServer(http.HandlerFunc(func(w http.ResponseWriter, req *http.Request) {
+		streamCh := make(chan *Stream)
+
+		w.WriteHeader(http.StatusSwitchingProtocols)
+
+		netconn, _, _ := w.(http.Hijacker).Hijack()
+		conn, _ := NewConnection(netconn, true)
+		go conn.Serve(func(s *Stream) {
+			s.SendReply(http.Header{}, false)
+			streamCh <- s
+		})
+
+		stream := <-streamCh
+		io.Copy(stream, stream)
+
+		closeChan := make(chan struct{})
+		go func() {
+			stream.Reset()
+			conn.Close()
+			close(closeChan)
+		}()
+
+		<-closeChan
+	}))
+
+	server.Start()
+	defer server.Close()
+
+	req, err := http.NewRequest("GET", server.URL, nil)
+	if err != nil {
+		t.Fatalf("Error creating request: %s", err)
+	}
+
+	rt := &roundTripper{}
+	client := &http.Client{Transport: rt}
+
+	_, err = client.Do(req)
+	if err != nil {
+		t.Fatalf("unexpected error from client.Do: %s", err)
+	}
+
+	conn, err := NewConnection(rt.conn, false)
+	go conn.Serve(NoOpStreamHandler)
+
+	stream, err := conn.CreateStream(http.Header{}, nil, false)
+	if err != nil {
+		t.Fatalf("error creating client stream: %s", err)
+	}
+
+	n, err := stream.Write([]byte("hello"))
+	if err != nil {
+		t.Fatalf("error writing to stream: %s", err)
+	}
+	if n != 5 {
+		t.Fatalf("Expected to write 5 bytes, but actually wrote %d", n)
+	}
+
+	b := make([]byte, 5)
+	n, err = stream.Read(b)
+	if err != nil {
+		t.Fatalf("error reading from stream: %s", err)
+	}
+	if n != 5 {
+		t.Fatalf("Expected to read 5 bytes, but actually read %d", n)
+	}
+	if e, a := "hello", string(b[0:n]); e != a {
+		t.Fatalf("expected '%s', got '%s'", e, a)
+	}
+
+	stream.Reset()
+	conn.Close()
 }
 
 var authenticated bool

--- a/Godeps/_workspace/src/github.com/docker/spdystream/stream.go
+++ b/Godeps/_workspace/src/github.com/docker/spdystream/stream.go
@@ -59,8 +59,6 @@ func (s *Stream) WriteData(data []byte, fin bool) error {
 		Data:     data,
 	}
 
-	s.conn.writeLock.Lock()
-	defer s.conn.writeLock.Unlock()
 	debugMessage("(%p) (%d) Writing data frame", s, s.streamId)
 	return s.conn.framer.WriteFrame(dataFrame)
 }
@@ -186,8 +184,6 @@ func (s *Stream) resetStream() error {
 		StreamId: s.streamId,
 		Status:   spdy.Cancel,
 	}
-	s.conn.writeLock.Lock()
-	defer s.conn.writeLock.Unlock()
 	return s.conn.framer.WriteFrame(resetFrame)
 }
 

